src/gui/servernotificationhandler.cpp

```diff
@@ -0,0 +1,100 @@
+ * Copyright (C) by Klaas Freitag <freitag@owncloud.com>
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; version 2 of the License.
+ *
+ * This program is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
+ * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License
+ * for more details.
+ */
+
+#include "servernotificationhandler.h"
+#include "accountstate.h"
+#include "capabilities.h"
+#include "json.h"
+#include "networkjobs.h"
+
+namespace OCC
+{
+
+ServerNotificationHandler::ServerNotificationHandler(QObject *parent)
+    : QObject(parent)
+{
+
+}
+
+void ServerNotificationHandler::slotFetchNotifications(AccountState *ptr)
+{
+    /* start the notification fetch job as well */
+    if( !ptr) {
+        return;
+    }
+
+    // check if the account has notifications enabled. If the capabilities are
+    // not yet valid, its assumed that notifications are available.
+    if( ptr->account() && ptr->account()->capabilities().isValid() ) {
+        if( ! ptr->account()->capabilities().notificationsAvailable() ) {
+            qDebug() << Q_FUNC_INFO << "Account" << ptr->account()->displayName() << "does not have notifications enabled.";
+            return;
+        }
+    }
+
+    // if the previous notification job has finished, start next.
+    _notificationJob = new JsonApiJob( ptr->account(), QLatin1String("ocs/v2.php/apps/notifications/api/v1/notifications"), this );
+    QObject::connect(_notificationJob.data(), SIGNAL(jsonReceived(QVariantMap, int)),
+                     this, SLOT(slotNotificationsReceived(QVariantMap, int)));
+    _notificationJob->setProperty("AccountStatePtr", QVariant::fromValue<AccountState*>(ptr));
+
+    _notificationJob->start();
+}
+
+void ServerNotificationHandler::slotNotificationsReceived(const QVariantMap& json, int statusCode)
+{
+    if( statusCode != 200 ) {
+        qDebug() << Q_FUNC_INFO << "Notifications failed with status code " << statusCode;
+        return;
+    }
+
+    auto notifies = json.value("ocs").toMap().value("data").toList();
+
+    AccountState* ai = qvariant_cast<AccountState*>(sender()->property("AccountStatePtr"));
+
+    ActivityList list;
+
+    foreach( auto element, notifies ) {
+        Activity a;
+        auto json   = element.toMap();
+        a._type     = Activity::NotificationType;
+        a._accName  = ai->account()->displayName();
+        a._id       = json.value("notification_id").toLongLong();
+        a._subject  = json.value("subject").toString();
+        a._message  = json.value("message").toString();
+        QString s   = json.value("link").toString();
+        if( !s.isEmpty() ) {
+            a._link     = QUrl(s);
+        }
+        a._dateTime = json.value("datetime").toDateTime();
+        a._dateTime.setTimeSpec(Qt::UTC);
```

Comment:
I'm not convinced that this setTimeSpec is a correct. Doesn't the ISO date string already denote the timezone (usually UTC)? And then this would either be a noop or, if the current timezone is different, change the dateTime to an incorrect value?

It does depend a lot on what the server communicates exactly.